### PR TITLE
Allow non-ligand, non-protein atoms to move beyond min_cutoff

### PR DESCRIPTION
**Pull request by proteneer**
During minimization, water molecules can spuriously move beyond the specified `min_cutoff`. This PR restricts the movement to that of only ligand and protein atoms. `InitialState` is now modified to include also `protein_idxs`.

### CODE REVIEW

File: timemachine/fe/rbfe.py [hunk 1]
```diff
@@ -357,8 +360,11 @@ def optimize_coordinates(initial_states, min_cutoff=0.7) -> List[NDArray]:
     for state, coords in zip(initial_states, all_xs):
         displacement_distances = jax_utils.distance_on_pairs(state.x0, coords, box=state.box0)
         if min_cutoff is not None:
+            # assert that ligand and protein atoms are not allowed to move more than min_cutoff
+            restricted_idxs = np.concatenate([state.ligand_idxs, state.protein_idxs])
+
             assert (
-                displacement_distances < min_cutoff
+                displacement_distances[restricted_idxs] < min_cutoff
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/fe/rbfe.py:367]: it's tricky - I need to carefully construct a system that can trigger a specific behavior in BFGS as I minimize. let me think about it.

File: timemachine/fe/rbfe.py [hunk 2]
```diff
@@ -357,8 +360,11 @@ def optimize_coordinates(initial_states, min_cutoff=0.7) -> List[NDArray]:
     for state, coords in zip(initial_states, all_xs):
         displacement_distances = jax_utils.distance_on_pairs(state.x0, coords, box=state.box0)
```
Comment (badisa) [inline timemachine/fe/rbfe.py:361]: Not for this PR, but could skip this entire loop (or at least the distance check) if `min_cutoff is None` which might speed things up in vacuum, though probably not much.

**Review by badisa — APPROVED** [review]

File: timemachine/fe/rbfe.py [hunk 3]
```diff
@@ -942,9 +950,11 @@ def run_edge_and_save_results(
                 [
                     f"{edge.mol_a_name} -> {edge.mol_b_name} (kJ/mol)",
                     f"exp_ddg {edge.metadata['exp_ddg']:.2f}" if "exp_ddg" in edge.metadata else "",
-                    f"fep_ddg {edge.metadata['fep_ddg']:.2f} +- {edge.metadata['fep_ddg_err']:.2f}"
-                    if "fep_ddg" in edge.metadata and "fep_ddg_err" in edge.metadata
-                    else "",
+                    (
+                        f"fep_ddg {edge.metadata['fep_ddg']:.2f} +- {edge.metadata['fep_ddg_err']:.2f}"
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/fe/rbfe.py:954]: yes - not sure why

File: timemachine/fe/free_energy.py [hunk 4]
```diff
@@ -120,6 +120,10 @@ class InitialState:
     ligand_idxs: NDArray
     protein_idxs: NDArray
 
+    def __post_init__(self):
+        assert self.ligand_idxs.dtype == np.int32 or self.ligand_idxs.dtype == np.int64
```
Comment (badisa) [inline timemachine/fe/free_energy.py:124]: Might also be good to add

`assert self.ligand_idxs.dtype == self.protein_idxs.dtype`

Just to be sure there isn't inconsistency between the two

File: timemachine/fe/rbfe.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (badisa) [reply timemachine/fe/rbfe.py:954]: I have noticed that black seems to behave differently under `git commit` for some files through precommit that aren't done by `precommit run --all-files`

File: timemachine/fe/free_energy.py (hunk 4 above)
Comment (proteneer) [reply timemachine/fe/free_energy.py:124]: This matters less - as long as both are int types, a concatenation will promote to the most precise value.

**Pull request merged**